### PR TITLE
[#14] [API] As a user, I can upload a CSV file of keywords

### LINKED ISSUE #14: [API] As a user, I can upload a CSV file of keywords
## Why

- An authenticated user can upload a CSV file
- On the home page, users can see a section for uploading files.

## Acceptance Criteria

Create a new route and endpoint.

### PR DESCRIPTION
**Pull request by vnntsu**
Closes #14 

## What happened 👀

Upload keywords file
- Create a new endpoint: `api/v1/keyword` with method POST.
- Validate authorization before uploading.

## Insight 📝

- Create `AuthPipeline` to ensure the request is called from an authorized user.
- Create `AuthErrorHandler` to show `unauthorized` error.

## Proof Of Work 📹

Screenshot 2022-12-26 at 22 24 41

### CODE REVIEW

File: lib/crawler_web/router.ex [hunk 1]
```diff
@@ -55,4 +59,12 @@ defmodule CrawlerWeb.Router do
       post("/log_in", UserSessionController, :create)
     end
   end
+
+  scope "/api", CrawlerWeb, as: :api do
+    pipe_through [:api, :api_auth]
+
+    scope "/v1", Api.V1, as: :v1 do
+      post("/keyword/upload", KeywordController, :create)
```
Comment (andyduong1920) [inline lib/crawler_web/router.ex]:
```suggestion
      post("/keyword", KeywordController, :create)
```

File: lib/crawler_web/controllers/api/v1/keyword_controller.ex [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+          conn,
+          gettext("One or more keywords are invalid! Allowed keyword length is %{min}-%{max}",
+            min: CSVParser.keyword_min_length(),
+            max: CSVParser.keyword_max_length()
+          )
+        )
+    end
+  end
+
+  defp show_error_flash_message_and_redirects_to_dasboard(conn, message) do
```
Comment (andyduong1920) [inline lib/crawler_web/controllers/api/v1/keyword_controller.ex]: This is not a flash message and we don't redirect the user here @vnntsu 

```suggestion
  defp render_error_message(conn, message) do
```

File: test/crawler_web/controllers/api/v1/keyword_controller_test.exs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+      list =
+        Enum.map(keywords, fn keyword ->
+          keyword.name
+        end)
+
+      assert list == ["one", "two", "three"]
+    end
+
+    test "given empty file, shows validation error", %{conn: conn} do
```
Comment (andyduong1920) [inline test/crawler_web/controllers/api/v1/keyword_controller_test.exs]: For API, in the test we prefer to write the result as `status`

```suggestion
    test "given empty file, returns 422 status", %{conn: conn} do
```

File: test/crawler_web/controllers/api/v1/keyword_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,148 @@
+defmodule CrawlerWeb.Api.V1.KeywordControllerTest do
+  use CrawlerWeb.ConnCase, async: true
+
+  alias Crawler.Keyword.Keywords
+
+  describe "POST /keyword/upload" do
+    test "given valid keyword csv file, creates keyword successfully", %{conn: conn} do
```
Comment (andyduong1920) [inline test/crawler_web/controllers/api/v1/keyword_controller_test.exs]:
```suggestion
    test "given valid keyword csv file, creates keywords and returns 200", %{conn: conn} do
```

File: lib/crawler_web/router.ex (hunk 1 above)
Comment (vnntsu) [reply lib/crawler_web/router.ex]: Fixed in [9c72659](https://github.com/vnntsu/elixir-ic/pull/37/commits/9c72659be8ef3893874a455f0fb0f09da067cb86)

**Review by andyduong1920 — APPROVED** [review]

**Review by Nihisil — APPROVED** [review]

**Pull request merged**